hoomd/pytest/test_gsd.py

```diff
@@ -0,0 +1,378 @@
+    snap.dihedrals.N = 1
+    snap.dihedrals.typeid[:] = [0]
+    snap.dihedrals.group[0] = [0, 1, 2, 3]
+
+    # impropers
+    snap.impropers.types = ['i1']
+    snap.impropers.N = 1
+    snap.impropers.typeid[:] = [0]
+    snap.impropers.group[0] = [3, 2, 1, 0]
+
+    # constraints
+    snap.constraints.N = 1
+    snap.constraints.group[0] = [0, 1]
+    snap.constraints.value[0] = 2.5
+
+    # special pairs
+    snap.pairs.types = ['p1', 'p2']
+    snap.pairs.N = 2
+    snap.pairs.typeid[:] = [0, 1]
+    snap.pairs.group[0] = [0, 1]
+    snap.pairs.group[1] = [2, 3]
+
+    return snap
+
+
+def test_write(simulation_factory, device,
+               hoomd_snapshot, tmp_path):
+
+    filename = tmp_path / "temporary_test_file.gsd"
+
+    sim = simulation_factory(hoomd_snapshot)
+    hoomd.write.GSD.write(state=sim.state, mode='wb',
+                          filename=str(filename))
+
+    with gsd.hoomd.open(name=filename, mode='wb') as traj:
+        assert_equivalent_snapshots(traj, hoomd_snapshot)
+
+
+def test_write_gsd_trigger(simulation_factory, device,
+                           hoomd_snapshot, tmp_path):
+
+    filename = tmp_path / "temporary_test_file.gsd"
+
+    sim = simulation_factory(hoomd_snapshot)
+
+    integrator = hoomd.md.Integrator(dt=0.005)
+    lj = hoomd.md.pair.LJ(nlist=hoomd.md.nlist.Cell(), r_cut=2.5)
+    lj.params[('p1', 'p1')] = {'sigma': 1, 'epsilon': 5e-200}
+    lj.params[('p1', 'p2')] = {'sigma': 1, 'epsilon': 5e-200}
+    lj.params[('p2', 'p2')] = {'sigma': 1, 'epsilon': 5e-200}
+    integrator.forces.append(lj)
+    langevin = hoomd.md.methods.Langevin(hoomd.filter.All(), kT=1, seed=1)
+    integrator.methods.append(langevin)
+    sim.operations.integrator = integrator
+
+    gsd_writer = hoomd.write.GSD(filename=filename,
+                                 trigger=hoomd.trigger.Periodic(1),
+                                 mode='wb', dynamic=['momentum'])
+    sim.operations.writers.append(gsd_writer)
+
+    snapshot_list = []
+    for _ in range(5):
+        sim.run(1)
+        snapshot_list.append(sim.state.snapshot)
+
+    with gsd.hoomd.open(name=filename, mode='rb') as traj:
+        for step in range(5):
+            assert_equivalent_snapshots(traj[step], snapshot_list[step])
+
+
+def test_write_gsd_mode(simulation_factory, device,
+                        hoomd_snapshot, tmp_path):
+
+    filename = tmp_path / "temporary_test_file.gsd"
+
+    sim = simulation_factory(hoomd_snapshot)
+
+    integrator = hoomd.md.Integrator(dt=0.005)
+    lj = hoomd.md.pair.LJ(nlist=hoomd.md.nlist.Cell(), r_cut=2.5)
+    lj.params[('p1', 'p1')] = {'sigma': 1, 'epsilon': 5e-200}
+    lj.params[('p1', 'p2')] = {'sigma': 1, 'epsilon': 5e-200}
+    lj.params[('p2', 'p2')] = {'sigma': 1, 'epsilon': 5e-200}
+    integrator.forces.append(lj)
+    langevin = hoomd.md.methods.Langevin(hoomd.filter.All(), kT=1, seed=1)
+    integrator.methods.append(langevin)
+    sim.operations.integrator = integrator
```

Comment:
This code is identical to the code above, could you create a fixture that is the integrator, or one that returns a simulation with the integrator added?

Reply:
Good idea. I realize now I copied too much